### PR TITLE
[HTML search] enable discovery of named caption ids

### LINKED ISSUE #11965: [HTML search] named captions have empty targets (anchors) in the titles index
### Describe the bug

As noted in #11961, we have some duplicate search results that appear in the frontend at the moment.

_Some_ of this is due to lack of JavaScript de-duplication of results, but it looks to me like we _also_ have duplication of entries in the `alltitles` part of the search index data that is built by the Python-side HTML/search-index builder code.

**Correction 2024-02-08**: it turns out that these entries are _not_ in fact duplicates.  By coincidence there is a table-of-contents (`toctree`) with a `caption` of `Get started` that emits an empty ID (because it has no `name` option configured).  I think we can improve the situation when a `name` _is_ configured - and #11966 is intended to do that.

### How to Reproduce

After self-building the Sphinx documentation from e976059fd6ab49b1c2784445d7095c9287897724 or a nearby commit, serve the contents of the built documentation via `python -m http.server -b 127.0.0.1` and then run a [query for `get started`](http://127.0.0.1:8000/search.html?q=get+started).

You should see two near-duplicate results at the top of the page; both link to the `index.html` page, one without an anchor, and one with the anchor/fragment `#get-started`.

image

Open the developer tools and inspect the contents of `Search._index.alltitles["Get started"]` in a browser JavaScript console.

image

Observe that the title heading `Get started` has two entries, both for the same file ID of `30`.  This seems to be a bug, and can contribute towards the presence of duplicate titles entries processed by the client.

### Environment Information

```text
Platform:              linux; (Linux-6.6.13-amd64-x86_64-with-glibc2.37)
Python version:        3.11.7 (main, Dec  8 2023, 14:22:46) [GCC 13.2.0])
Python implementation: CPython
Sphinx version:        7.2.6
Docutils version:      0.19
Jinja2 version:        3.1.3
Pygments version:      2.17.2
```

### Sphinx extensions

```python
N/A
```

### Additional context

Relates-to / discovered during investigation of #11961.

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- It seems that under some circumstances, items in the `alltitles` key of the JavaScript search index contain near-duplicates, where some items have an `id` (corresponding to an HTML anchor) and some do not.
- Given that all titles should (I think?) be hyperlinkable references, this changeset prevents title entries without IDs from being included in the search index.

### Detail
- ~~Skip `title` index entries when no IDs are present on the `title` node.~~
- ~~TBD: Work out why an equivalent entry _with_ an ID exists in the doctree.~~
- Allow the ID references for `caption` title nodes to be discovered from the parent of the `compact_paragraph` node that they are contained with.

### Relates
- May resolve #11965.

cc @wlach

### CODE REVIEW

**Comment by jayaddison** [comment]
> do not currently contain IDs

(unless the `:name:` option is specified, I think)

**Comment by picnixz** [comment]
Ah ! before doing anything with the search engine, I'm wondering whether #11707 is relevant or not.

**Comment by jayaddison** [comment]
> Ah ! before doing anything with the search engine, I'm wondering whether #11707 is relevant or not.

Thanks @picnixz - I'm not sure, but I don't _think_ that that's relevant in this case.  The problem here seems to be something to do with doctree nodes of type `nodes.title`, distinct from the singleton HTML `<title>`.

I'm narrowing in on the behaviour of the `:caption:` and `:name:` options on the `toctree` directive as the problem here, but I don't yet fully understand what's going on.

**Comment by jayaddison** [comment]
For example: applying the following patch breaks one of the tests, and that seems unexpected.

```diff
diff --git a/tests/roots/test-search/index.rst b/tests/roots/test-search/index.rst
index fc2298b5a..ccaed7c67 100644
--- a/tests/roots/test-search/index.rst
+++ b/tests/roots/test-search/index.rst
@@ -18,6 +18,7 @@ textinheading
 International
 
 .. toctree::
+   :caption: table-of-contents
 
    tocitem
 

```

**Comment by jayaddison** [comment]
Back to first principles: write a test case to illustrate the problem(s).

The problem here is that when the `caption` and `name` options are applied to an object, the resulting index entry has an empty target/ref, because the code that inspects the `nodes.title` object doesn't find the relevant ID  (the ID _does_ get created; it appears in the rendered HTML and can be `:ref:`'d.  so the problem appears to be mainly related to the `IndexBuilder`).

**Comment by jayaddison** [comment]
Figured it out, I think.  `caption` elements are a bit of a special-case; they do emit a `nodes.title` element, but that is nested within a [`compact_paragraph` element](https://github.com/sphinx-doc/sphinx/blob/e976059fd6ab49b1c2784445d7095c9287897724/sphinx/environment/adapters/toctree.py#L161).  So a lookup to find the `ids` won't succeed unless we check one level above in the node tree.

File: sphinx/search/__init__.py [hunk 1]
```diff
@@ -503,7 +503,9 @@ def _visit_nodes(node):
                 word_store.words.extend(split(node.astext()))
             elif isinstance(node, nodes.title):
                 title = node.astext()
-                ids = node.parent['ids']
+                ids = node.parent['ids'] or node.parent.parent['ids']
+                if not ids:
+                    return
```
Comment (jayaddison) [inline sphinx/search/__init__.py:508]: Note: there is also a (legacy?) equivalent [code block in `WordCollector`](https://github.com/sphinx-doc/sphinx/blob/e976059fd6ab49b1c2784445d7095c9287897724/sphinx/search/__init__.py#L232-L236).

**Comment by jayaddison** [comment]
Todo:

- [ ] Check (and test cover) whether `code-block` directive captions are also handled correctly.

**Comment by jayaddison** [comment]
> Todo:
> 
>     * [ ]  Check (and test cover) whether `code-block` directive captions are also handled correctly.

They're not - in fact, code snippets with named captions are omitted from the `title` index entirely at the moment.  I think that implies that it would be a separate feature request to add them / make them searchable (I've drafted some code to do that, but it seems worth keeping that separate from the bugfix here).

**Comment by jayaddison** [comment]
The bugreport that this pull request is linked to is confusing and has changed from its' initially-stated problem.  If it seems confusing to me -- I wrote it -- then I don't think it's likely to be understandable for future readers.

I'll close both and re-open a more precisely-defined bugreport after re-assessing the problem.

**Pull request closed without merging**